### PR TITLE
8326306: RISC-V: Re-structure MASM calls and jumps

### PR DESCRIPTION
**Pull request by robehn**
Hi, please consider.

We have code that directly use the asm for call/jumps instead masm.
Our masm have a bit odd naming, and we don't use 'proper' pseudoinstructions/mnemonics.
Suggested by [riscv-asm-manual](https://github.com/riscv-non-isa/riscv-asm-manual/tree/master)
```
j offset	jal x0, offset	Jump
jal offset	jal x1, offset	Jump and link
jr rs	        jalr x0, rs, 0	Jump register
jalr rs	        jalr x1, rs, 0	Jump and link register
ret	        jalr x0, x1, 0	Return from subroutine
call offset	auipc x1, offset[31:12]; jalr x1, x1, offset[11:0]	Call far-away subroutine	
tail offset	auipc x6, offset[31:12]; jalr x0, x6, offset[11:0]	Tail call far-away subroutine
```
But these can only be implemented like this if you have small enough application.
The fallback of these is to use GOT (your C compiler should place a copy of GOT every 2G so it's always reachable).
We don't have GOT, instead we materialize, so there is still differences between these and ours.

This patch:
- Tries to follow these suggested mappings as good we can.
- Make sure all jumps/calls go through MASM. (so we get control and can easily change for sites using a certain calling convention)
- To avoid confusion between MASM public/private methods and ASM methods and the mnemonics there are some renaming.
  E.g. the mnemonics jal means call offset, as we can't use that so there is no 'jal'.
- I enabled c.j, but right now we never generate it.
- As always the macro does no good and are legacy from when code base did not use templates. (also the x-macros screws up my IDE (vim+rtags))

I started down this path due to I have followup patch on top of this which removes trampoline in favor for load-n-jump.
(WIP: https://github.com/robehn/jdk/compare/jal-fixes...robehn:jdk:load-n-link?expand=1)
While looking into our calls it was a bit confusing, this helps. 

Done a couple of t1-3 slightly different version of this patch, and as part of the followup, no issues found. (VF2, qemu, LP4)
Re-running tests, had some last minute changes.

Thanks, Robbin

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8326306](https://bugs.openjdk.org/browse/JDK-8326306): RISC-V: Re-structure MASM calls and jumps (**Enhancement** - P4)

### Reviewers
 * [Fei Yang](https://openjdk.org/census#fyang) (@RealFYang - **Reviewer**)
 * [Ludovic Henry](https://openjdk.org/census#luhenry) (@luhenry - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18942/head:pull/18942` \
`$ git checkout pull/18942`

Update a local copy of the PR: \
`$ git checkout pull/18942` \
`$ git pull https://git.openjdk.org/jdk.git pull/18942/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18942`

View PR using the GUI difftool: \
`$ git pr show -t 18942`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18942.diff">https://git.openjdk.org/jdk/pull/18942.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18942#issuecomment-2076540837)

### CODE REVIEW

File: src/hotspot/cpu/riscv/macroAssembler_riscv.hpp [hunk 1]
```diff
@@ -689,6 +715,12 @@ class MacroAssembler: public Assembler {
     return x < (twoG - twoK) && x >= (-twoG - twoK);
   }
 
+  bool is_32bit_offset_from_codeache(int64_t x) {
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/macroAssembler_riscv.hpp:718]: Seems that this should be `is_32bit_offset_from_codecache`?

**Review by RealFYang — COMMENTED** [review]
Thanks for this cleanup! I am having a look.

File: src/hotspot/cpu/riscv/gc/shenandoah/shenandoahBarrierSetAssembler_riscv.cpp [hunk 2]
```diff
@@ -300,7 +300,7 @@ void ShenandoahBarrierSetAssembler::load_reference_barrier(MacroAssembler* masm,
     assert(!is_narrow, "phantom access cannot be narrow");
     target = CAST_FROM_FN_PTR(address, ShenandoahRuntime::load_reference_barrier_weak);
   }
-  __ call(target);
+  __ rt_call(target);
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/gc/shenandoah/shenandoahBarrierSetAssembler_riscv.cpp:303]: Question: does it make sense to replace `call` with `rt_call` when we are invoking the VM code (C++ code)? Here is what I see the difference of the two: `rt_call` emits code (`auipc` or `movptr`) depending on whether the destination could be found in code cache, while `call` depends on `is_32bit_offset_from_codeache`. So it's still possible for `call` to emit the short `auipc` code if not far even when the target is not there in the code cache like this case. But `rt_call` will always emit a long `movptr` sequence for this case.

Comment (robehn) [reply src/hotspot/cpu/riscv/gc/shenandoah/shenandoahBarrierSetAssembler_riscv.cpp:303]: I found some other places which uses plain calls to leaf, instead of cll_VM_leaf.
It seems like it's a guess that registers don't need push/pop, I don't think such speculation is good.
If it must be there we should have a argument to call_VM_leaf saying we don't want to push/pop.

x86 is this case uses the correct:
` __ call_VM_leaf(CAST_FROM_FN_PTR(address, ShenandoahRuntime::load_reference_barrier_weak), c_rarg0, c_rarg1);`

rt_call for non-code-cache do not need relocation, i.e. movptr.
So rt_call and call is not simply interchangeable, i.e. you now need relocation. 

Yes, we can probably do better here, but as the change is mv/li + jalr to movptr + jalr, there is no regression.
So improvement should be done outside of this PR.

Comment (robehn) [reply src/hotspot/cpu/riscv/gc/shenandoah/shenandoahBarrierSetAssembler_riscv.cpp:303]: Yes!

File: src/hotspot/cpu/riscv/stubGenerator_riscv.cpp [hunk 3]
```diff
@@ -5453,7 +5453,7 @@ static const int64_t right_3_bits = right_n_bits(3);
     }
     __ mv(c_rarg0, xthread);
     BLOCK_COMMENT("call runtime_entry");
-    __ call(runtime_entry);
+    __ rt_call(runtime_entry);
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/stubGenerator_riscv.cpp:5456]: I agree it's better to use `call_VM_leaf` for the Shenandoah cases. Then what about the changes in this file and templateInterpreterGenerator_riscv.cpp? Any reason to switch to `rt_call`?

Comment (robehn) [reply src/hotspot/cpu/riscv/stubGenerator_riscv.cpp:5456]: Old call():
```
    int32_t offset = 0;
    mv(temp, dest, offset); // =>li();
    jalr(x1, temp, offset);

```
To keep the sites the same (for non-code-cache calls)
New rt_call():
```
      movptr(tmp, target.target(), offset);
      Assembler::jalr(x1, tmp, offset);
```

Same here means absolute calls, no reloc required.
So I have tried to keep the calls the same. As you as we can optimize this by using reloc + la().

Comment (RealFYang) [reply src/hotspot/cpu/riscv/stubGenerator_riscv.cpp:5456]: Let me try to understand what you mean. So, are we going to remote the `relocate` call for non-code-cache call at [1] and further improve the `movptr` at [2] making use of `la`? This sounds interesting to me.

[1] https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/riscv/macroAssembler_riscv.cpp#L5031
[2] https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/riscv/macroAssembler_riscv.cpp#L5033

Comment (robehn) [reply src/hotspot/cpu/riscv/stubGenerator_riscv.cpp:5456]: I have not looked at it in detail.

As mnemonic for **call** is _auipc_ + _jalr_, in hotspot `la()` + `jalr()`.
So sites using call() for non-code-cache was changed to rt_call(), which gets us the same result as the old call().

Hence this patch 'tries' to keep the generate assembly the same.

Comment (robehn) [reply src/hotspot/cpu/riscv/stubGenerator_riscv.cpp:5456]: Make sense?

Comment (RealFYang) [reply src/hotspot/cpu/riscv/stubGenerator_riscv.cpp:5456]: I am still think about the possibility of unifying `call` and `rt_call`. Having both of them could be confusing to me (and new comers I guess). Like this add-on change:
[addon.diff.txt](https://github.com/openjdk/jdk/files/15164874/addon.diff.txt)
What do you think?

Comment (robehn) [reply src/hotspot/cpu/riscv/stubGenerator_riscv.cpp:5456]: Off today, I'll have a look tomorrow, thanks.

Comment (robehn) [reply src/hotspot/cpu/riscv/stubGenerator_riscv.cpp:5456]: Yes, we should use is_32bit_offset_from_codecache to use auipc as much as possible.
As there is much churn in this patch and our testing takes so long I was trying to keep assembly same to avoid bisecting down issues.

It's much better with just two calls, either rt call or VM leaf.

I'll test your patch, thanks.

Comment (robehn) [reply src/hotspot/cpu/riscv/stubGenerator_riscv.cpp:5456]: We still need relocates rt_call, not sure why you removed it.
It seem like we need two version of rt_call one with address and one with Address.
Then it seem like we could remove far_call as the rt_call would do the right thing.

I like your idea, and we should do that, but it seems like it's not trivial just to add to this patch.
Is there a reason we need to include such in changes in this PR?

**Comment by robehn** [comment]
My merge with master was hit by: https://bugs.openjdk.org/browse/JDK-8331546
Re-merge when it's fixed.

File: src/hotspot/cpu/riscv/stubGenerator_riscv.cpp (hunk 3 above)
Comment (RealFYang) [reply src/hotspot/cpu/riscv/stubGenerator_riscv.cpp:5453]: > We still need relocates rt_call, not sure why you removed it.

I removed the relocate because I am thinking it should be absolute calls in the else block of `MacroAssembler::rt_call` [1], so no reloc required, as you mentioned in your previous comments. 

[1] https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/riscv/macroAssembler_riscv.cpp#L5027

> It seem like we need two version of rt_call one with address and one with Address. Then it seem like we could remove far_call as the rt_call would do the right thing.
> 
> I like your idea, and we should do that, but it seems like it's not trivial just to add to this patch. Is there a reason we need to include such in changes in this PR?

On holiday this week, we can discuss further next week :-)

Comment (RealFYang) [reply src/hotspot/cpu/riscv/stubGenerator_riscv.cpp:5453]: > We still need relocates rt_call, not sure why you removed it. It seem like we need two version of rt_call one with address and one with Address. Then it seem like we could remove far_call as the rt_call would do the right thing.

I guess maybe it's easier to only unify `call` and `rt_call` as a first step? And I prefer to keep `far_call` and `far_jump` as I think they are kind of different from rt_call. These two only handle the case when target is within code cache [1][2] as compared to rt_call which handles both code-cache and non-code-cache targets.

[1] https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/riscv/macroAssembler_riscv.cpp#L3173
[2] https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/riscv/macroAssembler_riscv.cpp#L3188

> I like your idea, and we should do that, but it seems like it's not trivial just to add to this patch. Is there a reason we need to include such in changes in this PR?

Another concern which also leads me to consider unifying `call` and `rt_call` is possible performance issues. I see some `call` are changed into `rt_call` like in file stubGenerator_riscv.cpp and templateInterpreterGenerator_riscv.cpp. As I mentioned in my previous comments, `rt_call` would emit a fixed-size `movptr` sequence (6 uncompressed instructions) for these call sites which invokes some C++ VM functions. But it's still possible for the original `call` to emit a more simpler auipc + jalr depending on the `is_32bit_offset_from_codecache` check in `la`, right? Hopefully, we can get rid of this issue with my proposed change.

Comment (robehn) [reply src/hotspot/cpu/riscv/stubGenerator_riscv.cpp:5453]: Yes, sorry I didn't explain that:
The cases in stubGenerator are all slow-path so we don't care much about performance.
Calling:
- MacroAssembler::debug64 debug
- JavaThread::check_special_condition_for_native_trans when suspend flag is set
- SharedRuntime::reguard_yellow_pages regaurd pages => syscall
- Interpreter::trace_code(t->tos_in()) Calling a interpreter trace method for the bytcode

Comment (RealFYang) [reply src/hotspot/cpu/riscv/stubGenerator_riscv.cpp:5453]: All right. What about the one in `MacroAssembler::call_VM_leaf_base`? The `call` there is replaced with movptr+jalr. Why not use `rt_call` here like the other places? Then we won't lose it when we eliminate `call` finally.

Comment (robehn) [reply src/hotspot/cpu/riscv/stubGenerator_riscv.cpp:5453]: Yes, you right the old:
call() = mv(tmp, adr) => li(adr) + jalr()

li(adr) is better than movptr()

I'll fix thanks!

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp [hunk 4]
```diff
@@ -859,88 +847,113 @@ void MacroAssembler::li(Register Rd, int64_t imm) {
...
+}
 
-  INSN(j,   x0);
-  INSN(jal, x1);
+void MacroAssembler::call(const address dest, Register temp) {
+  assert_cond(dest != nullptr);
+  assert(temp != noreg, "expecting a register");
+  int32_t offset = 0;
+  la(temp, dest, offset);
+  Assembler::jalr(x1, temp, offset);
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:938]: Hi, Did you check the possible impact on performance of this change?

old: call() => mv(tmp, adr) => li(tmp, adr)
new: call() => la(tmp, adr) => auipc/movptr(tmp, adr)

So the new call() could emit a fixed-sized movptr depending on the `is_32bit_offset_from_codecache` check which should be slower than a li(tmp, adr).

Comment (robehn) [reply src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:938]: I checked that we use the use call with a code cache address for the places we care about such as:
I.e. JNI FastGetField and far_call()

But I now see I was mistaken regarding the interpreter, it seems like we don't have all the math implemented, so we do calls to SharedRuntime there, I'll fix those.

Thanks!

Comment (robehn) [reply src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:941]: Great, thanks.

I fixed it by using li() in la() when we wan't use auipc, running tests.

Comment (robehn) [reply src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:938]: That was bad idea.

Looking into something else.

Comment (robehn) [reply src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:938]: It's a bit annoying, since if we are about performance in the interpreter we should implement the math stubs, e.g. StubRoutines::dcos.

Comment (robehn) [reply src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:938]: I don't find a nice way to get li for them.
https://github.com/openjdk/jdk/blob/3b8227ba24c7bc05a8ea23801e3816e8fc80de4e/src/hotspot/cpu/riscv/templateInterpreterGenerator_riscv.cpp#L272

Any suggestions?

As I said, if we want to be performant here we should implement the "StubRoutines::dpow()".

What you think?

Comment (robehn) [reply src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:942]: As I can't fix this with a few changes, and I don't want to mess up this PR.
I added check for li directly in the call.
Sanity tested with -XX:ReservedCodeCacheSize=2047M

File: src/hotspot/cpu/riscv/jniFastGetField_riscv.cpp [hunk 5]
```diff
@@ -175,9 +175,7 @@ address JNI_FastGetField::generate_fast_get_int_field0(BasicType type) {
     __ enter();
     ExternalAddress target(slow_case_addr);
     __ relocate(target.rspec(), [&] {
-      int32_t offset;
-      __ la(t0, target.target(), offset);
-      __ jalr(x1, t0, offset);
+      __ call(target.target());
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/jniFastGetField_riscv.cpp:178]: Should we revert this change after your last commit? As I think call is now not necessarily la + jalr.

Comment (robehn) [reply src/hotspot/cpu/riscv/jniFastGetField_riscv.cpp:178]: The addresses should come from JNI_FastGetField::generate_fast_get_XXX_field0.
So slow_case_addr is not really an ExternalAddress.
This means call will always do auipc + jalr as you know (intra code cache).

Comment (RealFYang) [reply src/hotspot/cpu/riscv/jniFastGetField_riscv.cpp:178]: Hmm... So I did a quick try on linux-aarch64 invoking `CodeCache::contains` on slow_case_addr and the result is false. Anything I missed?

Comment (robehn) [reply src/hotspot/cpu/riscv/jniFastGetField_riscv.cpp:178]: The JNI_FastGetField::generate_fast_get_XXX_field0 write the code in a CodeBuffer, this where slow_case_addr points to.

As I test with ReservedCodeCacheSize=2047M if we generated an li() relocation would fail. (there is usually around 120 MB between them)

Added your assert, it passes also.
```
     ExternalAddress target(slow_case_addr);
+    assert(CodeCache::contains(slow_case_addr), "Must be");
     __ relocate(target.rspec(), [&] {

```

I guess you are running on apple, the code concerning "static_fast_get_field_wrapper" smells.
Maybe you found a bug here.

Comment (RealFYang) [reply src/hotspot/cpu/riscv/jniFastGetField_riscv.cpp:178]: Try this: `make test TEST="runtime/jni/FastGetField/FastGetField.java"`
The assertion failed on both linux-aarch64 and linux-riscv64 platforms.

Comment (robehn) [reply src/hotspot/cpu/riscv/jniFastGetField_riscv.cpp:178]: As you turn off "fast JNI" with those option I think it's fine to use rt_call, which would do movptr in this case.
If you wanted fast JNI you should,'t turn it off.

But I can revert to la+jalr if you feel strongly about it?

Comment (RealFYang) [reply src/hotspot/cpu/riscv/jniFastGetField_riscv.cpp:178]: > But I can revert to la+jalr if you feel strongly about it?

Yeah, I would suggest revert to the original la+jalr for now and think more about it.

Comment (robehn) [reply src/hotspot/cpu/riscv/jniFastGetField_riscv.cpp:178]: Yes, let's finish this PR and make other wanted changes in other PR's :)

File: src/hotspot/cpu/riscv/assembler_riscv.hpp [hunk 6]
```diff
@@ -2813,24 +2812,35 @@ enum Nf {
...
+    _jalr(Rd, Rs, offset);
   }
 
-  INSN(jalr);
+  void jal(Register Rd, const int32_t offset) {
+    /* jal -> c.j, note c.jal is RV32C only */
+    if (do_compress() &&
+        Rd == x0 &&
+        is_simm12(offset) && ((offset % 2) == 0)) {
+      c_j(offset);
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/assembler_riscv.hpp:2836]: Is RV32C-only instructions usable for our case? Or will this if block be test covered?

**Review by RealFYang — COMMENTED** [review]
Thanks for the quick update. Two minor comments remain :-)

File: src/hotspot/cpu/riscv/macroAssembler_riscv.hpp [hunk 7]
```diff
@@ -735,12 +767,12 @@ class MacroAssembler: public Assembler {
   typedef void (MacroAssembler::* jal_jalr_insn)(Register Rt, address dest);
   typedef void (MacroAssembler::* load_insn_by_temp)(Register Rt, address dest, Register temp);
 
-  void wrap_label(Register r, Label &L, Register t, load_insn_by_temp insn);
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/macroAssembler_riscv.hpp:738]: Better to remove the `load_insn_by_temp` declaration from file macroAssembler_riscv.hpp as it is not used anywhere after this change.

`typedef void (MacroAssembler::* load_insn_by_temp)(Register Rt, address dest, Register temp);`

File: src/hotspot/cpu/riscv/macroAssembler_riscv.hpp [hunk 8]
```diff
@@ -689,6 +715,12 @@ class MacroAssembler: public Assembler {
     return x < (twoG - twoK) && x >= (-twoG - twoK);
   }
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/macroAssembler_riscv.hpp:717]: Can you add a code comment for this function to make it easier to understand? Maybe something like:
`Ensure that the auipc can reach the destination at x from anywhere within the code cache so that if it is relocated we know it will still reach.`

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp [hunk 9]
```diff
@@ -641,14 +641,16 @@ void MacroAssembler::emit_static_call_stub() {
...
 }
 
 void MacroAssembler::call_VM_leaf_base(address entry_point,
                                        int number_of_arguments,
                                        Label *retaddr) {
+  int32_t offset = 0;
   push_reg(RegSet::of(t0, xmethod), sp);   // push << t0 & xmethod >> to sp
-  call(entry_point);
+  mv(t0, entry_point, offset);
+  Assembler::jalr(x1, t0, offset);
```
[the remark this answers is not in the document]
Comment (robehn) [reply src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:653]: Fixed

File: src/hotspot/cpu/riscv/assembler_riscv.hpp (hunk 6 above)
Comment (robehn) [reply src/hotspot/cpu/riscv/assembler_riscv.hpp:2836]: There are two instructions using the CJ format: c.j and c.jal.
- c.j is already in our assembler.
- c.jal is not in our assembler and is RV32C only.

000000000000064e <c_j>:
     64e: 11 a0        	j	0x652 <c_j+0x4>  // 0xa011 is c.j 4
     650: 2a 85        	mv	a0, a0
     652: 82 80        	ret

Works fine on both VF2 and qemu-rv64.

The comment was just saying that we can't try to map to c.jal since it's RV32C.

The instruction c.j have the same test coverage as before (as we already have it in assembler).
The toogle to C I think is untested as it only can be generated for short backwards branches in non-relocated code.

As the comment is obvious confusing, suggestions to change it?

Comment (RealFYang) [reply src/hotspot/cpu/riscv/assembler_riscv.hpp:2836]: Uh, I see. I think I mis-read the comment. I am OK with the current shape.

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp [hunk 10]
```diff
@@ -859,88 +847,113 @@ void MacroAssembler::li(Register Rd, int64_t imm) {
...
-#undef INSN
+void MacroAssembler::jump_link(const address dest, Register temp) {
+  assert_cond(dest != nullptr);
+  int64_t distance = dest - pc();
+  if (is_simm21(distance) && ((distance % 2) == 0)) {
+    Assembler::jal(x1, distance);
+  } else {
+    assert(temp != noreg && temp != x0, "expecting a register");
+    int32_t offset = 0;
+    movptr(temp, dest, offset);
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:858]: One more question: will `la(temp, dest, offset)` work here? I see you've changed to use `la` for `MacroAssembler::j(const address dest, Register temp)` instead of the `movptr` used before Seems we can do similar change here as well.

File: src/hotspot/cpu/riscv/macroAssembler_riscv.hpp (hunk 7 above)
Comment (robehn) [reply src/hotspot/cpu/riscv/macroAssembler_riscv.hpp:738]: Fixed

File: src/hotspot/cpu/riscv/macroAssembler_riscv.hpp (hunk 8 above)
Comment (robehn) [reply src/hotspot/cpu/riscv/macroAssembler_riscv.hpp:717]: Fixed

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp (hunk 10 above)
Comment (robehn) [reply src/hotspot/cpu/riscv/macroAssembler_riscv.cpp]: Jump can 'only' go to code cache (if someone jumps to C/C++ method we would be in trouble).
Therefore I changed to la(), otherwise I tried to keep generate code same:ish.

I don't see any reason for why we can't use it for those cases.

I think any movptr in a compressible region should be la().

Sanity tested with RCC 2047 and unset.

**Review by RealFYang — APPROVED** [review]
Updated change looks great! Thank you.

**Comment by robehn** [comment]
> Updated change looks great! Thank you.

Thanks for sticking with it, and thanks for a good review!

**Comment by RealFYang** [comment]
So now `MacroAssembler::rt_call` is effectively this after this PR change:
```
void MacroAssembler::rt_call(address dest, Register tmp) {
  RuntimeAddress target(dest);
  relocate(target.rspec(), [&] {
    int32_t offset;
    la(tmp, target.target(), offset);
    jalr(tmp, offset);
  });
}
```

Here are some of my thoughts on the next step/cleanup for reference:

1. Since there are only a few uses of `MacroAssembler::call`, we might want to remove this function
replacing its callsites with `MacroAssembler::rt_call`. This would help avoid possible confusion
between `call` and `rt_call`. I don't think the relocation added by `rt_call` would make a difference.

2. For `MacroAssembler::far_call` and `MacroAssembler::far_jump`, I would suggest we use direct `auipc`
instead of `la` for them as the destination is expected to be in code cache. This will help distinguish
these two functions from `MacroAssembler::rt_call`.

3. As we discussed, we also want to fix Shenandoah related callsites to `call_VM_leaf`. And let call_VM_leaf to use `rt_call` so that we emit `auipc` when possible.

**Comment by robehn** [comment]
Yes, in general agreed.

One issue is that RV MASM often have lower-case address, where other platforms use Address, i.e. AddressLiteral on x86.
I think we should use upper case Address as then we can check for relocation and produce all different code patterns with same function, li/auipc/movptr.

While looking at this it seem like we have some funny things:
```
src/hotspot/cpu/riscv/c1_CodeStubs_riscv.cpp:  __ rt_call(Runtime1::entry_for(stub_id), ra);
src/hotspot/cpu/riscv/c1_CodeStubs_riscv.cpp:  __ far_call(RuntimeAddress(Runtime1::entry_for(_stub_id)));

```

Note that 'mv reg, imm' is not a recognized mnemonic (should be 'li reg, imm').

**Comment by RealFYang** [comment]
> RuntimeAddress target(dest);

Yeah, I know what you mean. As you see, the `address` passed to `rt_call` is used to construct a `RuntimeAddress` which is also an `Address`. It seems that it will be more consistent to build a `RuntimeAddress` and pass that to `rt_call`, which is similar with others like `far_call`.

**Comment by robehn** [comment]
Passed t1 with RCC 2047M, t1 with default RCC rolling on fine. (we have very noise t1 at the moment)

Any other takers?

**Review by luhenry — APPROVED** [review]

**Comment by robehn** [comment]
Thank you @luhenry.

I did a full t1-t3 (t1 also with 2047) (VF2) I have seen no issues with this, I'll integrate later today.
